File: EspressifProvision/EspressifProvision/provision/crypto/AESDecryptor.m
//
//  Encrpyt.m
//  EspressifProvision
//
//  Created by Vikas Chandra on 15/05/19.
//  Copyright © 2019 Espressif. All rights reserved.
//

#import "AESDecryptor.h"
@interface AESDecryptor() {
    CCCryptorRef cryptor;
}
@end

@implementation AESDecryptor

-(id)initWithKey:(NSData *)key andIV:(NSData *)iv {
    cryptor = NULL;
    CCCryptorStatus ccStatus = 0;
    
    ccStatus = CCCryptorCreateWithMode(kCCEncrypt, kCCModeCTR, kCCAlgorithmAES,
                                       ccNoPadding,
                                       iv.bytes, key.bytes,
                                       kCCKeySizeAES256,
                                       NULL, 0, 0, // tweak XTS mode, numRounds
                                       kCCModeOptionCTR_BE, // CCModeOptions
                                       &cryptor);
    return self;
}
- (NSData *)cryptData:(NSData *)dataIn
            operation:(CCOperation)operation  // kCC Encrypt, Decrypt
                 mode:(CCMode)mode            // kCCMode ECB, CBC, CFB, CTR, OFB, RC4, CFB8
            algorithm:(CCAlgorithm)algorithm  // CCAlgorithm AES DES, 3DES, CAST, RC4, RC2, Blowfish
              padding:(CCPadding)padding      // cc NoPadding, PKCS7Padding
            keyLength:(size_t)keyLength       // kCCKeySizeAES 128, 192, 256
                   iv:(NSData *)iv            // CBC, CFB, CFB8, OFB, CTR
                  key:(NSData *)key
                error:(NSError **)error
{
    if (key.length != keyLength) {
        NSLog(@"CCCryptorArgument key.length: %lu != keyLength: %zu", (unsigned long)key.length, keyLength);
        if (error) {
            *error = [NSError errorWithDomain:@"kArgumentError key length" code:key.length userInfo:nil];
        }
        return nil;
    }
    
    size_t dataOutMoved = 0;
    size_t dataOutMovedTotal = 0;
    CCCryptorStatus ccStatus = 0;
    
    size_t dataOutLength = CCCryptorGetOutputLength(cryptor, dataIn.length, true);
    NSMutableData *dataOut = [NSMutableData dataWithLength:dataOutLength];
    char *dataOutPointer = (char *)dataOut.mutableBytes;
    
    ccStatus = CCCryptorUpdate(cryptor,
                               dataIn.bytes, dataIn.length,
                               dataOutPointer, dataOutLength,
                               &dataOutMoved);
    dataOutMovedTotal += dataOutMoved;
    
    if (ccStatus != kCCSuccess) {
        NSLog(@"CCCryptorUpdate status: %d", ccStatus);
        if (error) {
            *error = [NSError errorWithDomain:@"kUpdateError" code:ccStatus userInfo:nil];
        }
        CCCryptorRelease(cryptor);
        return nil;
    }
    
    ccStatus = CCCryptorFinal(cryptor,
                              dataOutPointer + dataOutMoved, dataOutLength - dataOutMoved,
                              &dataOutMoved);
    if (ccStatus != kCCSuccess) {
        NSLog(@"CCCryptorFinal status: %d", ccStatus);
        if (error) {
            *error = [NSError errorWithDomain:@"kFinalError" code:ccStatus userInfo:nil];
        }
        CCCryptorRelease(cryptor);
        return nil;
    }
    
    dataOutMovedTotal += dataOutMoved;
    dataOut.length = dataOutMovedTotal;
    
    return dataOut;
}
@end

File: Sources/ESPProvision/Crypto/AESDecryptor.swift
//
//  AESDecryptor.swift
//  ESPProvision
//
//  Created by Martin Zarzeczny on 25.02.22.
//

import Foundation
import CommonCrypto

public class AESDecryptor {

    private var cryptor: CCCryptorRef? = nil

    init(with key: Data, and iv: Data) {
        var ccStatus: CCCryptorStatus = 0;

        ccStatus = CCCryptorCreateWithMode(CCOperation(kCCEncrypt), CCMode(kCCModeCTR), CCAlgorithm(kCCAlgorithmAES),
                                           CCPadding(ccNoPadding),
                                           iv.bytes, key.bytes,
                                           kCCKeySizeAES256,
                                           nil, 0, 0,
                                           CCModeOptions(kCCModeOptionCTR_BE),
                                           &cryptor)
    }

    func cryptData(dataIn: Data, operation: CCOperation, mode: CCMode, algorithm: CCAlgorithm, padding: CCPadding, keyLength: size_t, iv: Data, key: Data, error: inout NSError?) -> Data? {
        if key.count != keyLength {
            ESPLog.log("CCCryptorArgument key.length: \(key.count) != keyLength: \(keyLength)");
            if (error != nil) {
                error = NSError(domain: "kArgumentError key length", code: key.count, userInfo: nil)
            }
            return nil;
        }

        var dataOutMoved: size_t = 0
        var dataOutMovedTotal: size_t = 0
        var ccStatus: CCCryptorStatus = 0

        var accumulator : [UInt8] = []

        let dataOutLength: size_t = CCCryptorGetOutputLength(cryptor, dataIn.count, true);
        var dataOut = Array<UInt8>(repeating: 0, count:Int(dataOutLength))

        ccStatus = CCCryptorUpdate(cryptor,
                                   dataIn.bytes, dataIn.count,
                                   &dataOut, dataOutLength,
                                   &dataOutMoved);
        dataOutMovedTotal += dataOutMoved;

        if (ccStatus != kCCSuccess) {
            ESPLog.log("CCCryptorUpdate status: \(ccStatus)")
            if (error != nil) {
                error = NSError(domain: "kUpdateError", code: Int(ccStatus), userInfo: nil)
            }
            CCCryptorRelease(cryptor);
            return nil;
        }

        accumulator += dataOut[0..<Int(dataOutMoved)]
        ccStatus = CCCryptorFinal(cryptor, &dataOut, dataOutLength, &dataOutMoved)

        if (ccStatus != kCCSuccess) {
            ESPLog.log("CCCryptorFinal status: \(ccStatus)");
            if (error != nil) {
                error = NSError(domain: "kFinalError", code: Int(ccStatus), userInfo: nil)
            }
            CCCryptorRelease(cryptor);
            return nil;
        }

        dataOutMovedTotal += dataOutMoved;
        accumulator += dataOut[0..<Int(dataOutMoved)]

        return Data(accumulator)
    }
}
